include/zephyr/drivers/usb/udc_buf.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Buffers for USB device support
+ */
+
+#ifndef ZEPHYR_INCLUDE_UDC_BUF_H
+#define ZEPHYR_INCLUDE_UDC_BUF_H
+
+#include <zephyr/net/buf.h>
+
+#if defined(CONFIG_DCACHE) && !defined(CONFIG_NOCACHE_MEMORY)
+/*
+ * Here we try to get DMA-safe buffers, but we lack a consistent source of
+ * information about data cache properties, such as line cache size, and a
+ * consistent source of information about what part of memory is DMA'able.
+ * For now, we simply assume that all available memory is DMA'able and use
+ * Kconfig option DCACHE_LINE_SIZE for alignment and granularity.
+ */
+#define UDC_BUF_ALIGN		CONFIG_DCACHE_LINE_SIZE
+#define UDC_BUF_GRANULARITY	CONFIG_DCACHE_LINE_SIZE
+#else
+/* Defaults alignment and granularity if there is not data cache */
+#define UDC_BUF_ALIGN		sizeof(void *)
+#define UDC_BUF_GRANULARITY	sizeof(void *)
+#endif
+
+#define UDC_K_HEAP_DEFINE_NO_CACHE(name, size)						\
+	K_HEAP_DEFINE_NOCACHE(name, size)
+
+#define UDC_K_HEAP_DEFINE_DMAABLE(name, size)						\
+	K_HEAP_DEFINE(name, size)
+
+#define UDC_K_HEAP_DEFINE(name, size)							\
+	COND_CODE_1(CONFIG_NOCACHE_MEMORY,						\
+		    (UDC_K_HEAP_DEFINE_NO_CACHE(name, size)),				\
+		    (UDC_K_HEAP_DEFINE_DMAABLE(name, size)))
+
+extern const struct net_buf_data_cb net_buf_dma_cb;
+
+/**
+ * @brief Buffer API for USB device support
+ * @defgroup udc_buf Experimental buffer API for USB device support
+ * @ingroup usb
+ * @{
+ */
+
+/**
+ * @brief Define a new pool for UDC buffers based on fixed-size data
+ *
+ * This macro is similar to `NET_BUF_POOL_VAR_DEFINE`, but provides buffers
+ * with proper alignment and granularity.
+ *
+ * @see NET_BUF_POOL_VAR_DEFINE
+ *
+ * @param pname      Name of the pool variable.
+ * @param count      Number of buffers in the pool.
+ * @param size       Maximum data payload per buffer.
+ * @param ud_size    User data space to reserve per buffer.
+ * @param fdestroy   Optional destroy callback when buffer is freed.
+ */
+#define UDC_BUF_POOL_VAR_DEFINE(pname, count, size, ud_size, fdestroy)			\
```

Comment:
Is this replacement for `NET_BUF_POOL_DEFINE`? If yes, then it needs to be used in `usbd_uac2.c` and `usbd_hid_macros.h`

It would be good to have alignment assert in `uac2_buf_alloc()` and `hid_buf_alloc_ext()`, e.g. `__ASSERT(IS_ALIGNED(data, UDC_BUF_ALIGN), "Application provided unaligned buffer")`. The granurality is unfortunately not possible to be checked and has to rely on application developer providing appropriately padded buffers.